wallet-new/src/Cardano/Wallet/Kernel/Submission.hs

```diff
@@ -0,0 +1,431 @@
+            let newState = ws & over (wsState . wssSchedule) (const schedule')
+                              . over (wsState . wssCurrentSlot) (mapSlot succ)
+            return (toPrune , newState)
+
+data SchedulingError =
+    LoopDetected Pending
+    -- ^ The transactions in this 'Pending' set forms a cycle and they
+    -- couldn't be top-sorted.
+
+instance Exception SchedulingError
+
+instance Show SchedulingError where
+    show (LoopDetected pending) = toString $ sformat ("LoopDetected " % F.build) pending
+
+-- | Returns a set of 'Pending' transactions which are due in the given
+-- 'Slot'.
+-- TODO(adn) Document this.
+--
+dueThisSlot :: Slot
+            -- ^ The current 'Slot'.
+            -> WalletSubmission m
+            -- ^ The 'WalletSubmissionState'.
+            -> Either SchedulingError ([ScheduledTx], Schedule)
+            -- ^ An error if no schedule can be produced, or the scheduled
+            -- transactions together with the new, updated 'Schedule'.
+dueThisSlot currentSlot ws =
+    let localPending = ws ^. wsState . wssPendingSet
+        (dependent, independent) = partitionScheduledTx localPending candidates
+        topSorted  = topsortTxs toTx independent
+    in case topSorted of
+            Nothing     -> Left (LoopDetected localPending)
+            Just sorted ->
+                let newSchedule = schedule { _ssUnsentNursery = dependent }
+                in Right (sorted, purgeSlot currentSlot newSchedule)
+    where
+        candidates :: [ScheduledTx]
+        candidates =
+            let due = fromMaybe mempty (IntMap.lookup (castSlot currentSlot) (_ssScheduled schedule))
+            in due `mappend` _ssUnsentNursery schedule
+
+        -- The current 'Schedule'.
+        schedule :: Schedule
+        schedule = ws ^. getSchedule
+
+        toTx :: ScheduledTx -> WithHash Core.Tx
+        toTx (ScheduledTx txId txAux _) =  WithHash (Core.taTx txAux) txId
+
+
+partitionScheduledTx :: Pending -> [ScheduledTx] -> ([ScheduledTx], [ScheduledTx])
+partitionScheduledTx pending xs = List.partition dependsOnPendingTx xs
+    where
+        -- | A 'ScheduledTx' is @not@ independent and should not be sent
+        -- over the wire if any of the inputs it consumes are mentioned in
+        -- the 'Pending' set.
+        dependsOnPendingTx :: ScheduledTx -> Bool
+        dependsOnPendingTx (ScheduledTx _ txAux _) =
+            let txs = pending ^. pendingTransactions . fromDb
+                inputs = List.foldl' updateFn mempty $ (Core.taTx txAux) ^. Core.txInputs . to NonEmpty.toList
+            in any (isJust . flip M.lookup txs) inputs
+
+        updateFn :: [Core.TxId] -> Core.TxIn -> [Core.TxId]
+        updateFn acc (Core.TxInUnknown _ _)   = acc
+        updateFn acc (Core.TxInUtxo txHash _) = txHash : acc
+
+
+-- | Overrides the 'Schedule' with an input transaction and a
+-- custom 'Schedule'. Useful to force dispatching, especially in tests.
```

Comment:
I know I suggested this name but reading through this again, `override` is a bit of a misnomer actually no? It suggests that we replace something, but in fact we're appending. What do you think?

Reply:
I am a bit on the fence -- although I agree that this is just appending (so maybe `consSchedule` or `snocSchedule` might be a more suitable name) the interpretation that this is "overriding" a `Schedule` is also correct -- we are, in a sense, hijacking the Schedule (which handling would be normally internal or part of the resubmission function logic
).


No strong feelings, I'm open to both solutions.

Reply:
An alternative might be to parametrise our function with a `(Schedule -> Schedule)` override function.